flips/20220715-cadence-purity-analysis.md

```diff
@@ -0,0 +1,211 @@
+        return a.length
+    }
+    ```
+
+    Cases like this will have to be rejected statically for safety. Field reads, of course, will remain valid, as they do not modify any state. 
+
+* Writing to storage. This includes operations like `save` and `load` (since if a resource is loaded it will be moved out of storage), as well as `unlink`, but
+not operations like `copy`, `borrow`, `type`, or `getCapability`.
+
+* Writing to account state. This includes operations like adding, updating or removing a contract and adding, updating or removing account keys.
+
+* Logging data (calling the `log` function)
+
+* Emitting events
+
+* Mutating a reference. Any writes to references will have to be rejected, as it is generally not possible to know what value is being referenced. Consider code like the following:
+
+    ```cadence
+    view fun foo(a: [Int], b: bool): Int {
+        let c: [Int] = &[] as &[Int]
+        if b {
+            c = &a as &[Int]
+        }
+        c.append(4) // it is impossible to tell statically whether `a` is receiving the write here
+        return a.length
+    }
+    ```
+
+Any functions declared to be `view` will enforce that none of these operations occur in their bodies. If such an operation does occur, then the checker will report
+an error identifying the impure operation. Impure function declarations have no additional typechecking behavior; any operations are permitted.
+
+Note that the current design allows both emitting events and logging data from within a view function. These are traditionally considered side effects, but 
+it is perhaps worth some discussion on whether or not these should be allowed in `view` functions.  
+
+Purity also interacts covariantly with function subtyping: `view` functions are a subtype of impure functions. So, the following declarations would typecheck:
+
+```cadence
+    let a: (view (): Void) = view fun() {}
+    let b: ((): Void) = view fun() {}
+    let c: ((): Void) = fun() {}
+    let d: (((view (): Void)): Void) = fun foo(x:((): Void)) {} // contravariance
+```
+
+while these would not: 
+
+
+```cadence
+    let x: (view (): Void) = fun() {}
+    let y: ((((): Void)): Void) = fun foo(f:(view (): Void)) {} // contravariance
+```
+
+Once purity is enforced in functions with `view` annotations, in order to require it in function conditions we can 
+simply treat pre-conditions and post-conditions as view contexts as if they had `view` annotations themselves. 
+
+### Drawbacks
+
+This adds additional complexity to the langauge; users will need to think about the purity of their functions and how they interact with each other in order 
+to use the new `view` annotations. However, these annotations are unlikely to be frequently required; they will only be necessary for functions that are 
+called from conditions (and functions called by those functions). Users will not need to interact with this system most of the time. 
```

Comment:
```suggestion
The proposed feature adds additional complexity to the language; 
users will need to think about the side-effects of their functions and how they interact with each other in order  to use the new `view` annotation.
However, this annotation is unlikely to be frequently required; it will only be necessary for functions that are 
called from conditions (and functions called by those functions). 
Users will not need to interact with this language feature most of the time. 
```